chromadb/api/collection_configuration.py

```diff
@@ -0,0 +1,817 @@
+    Space,
+    Metadata,
+    CollectionMetadata,
+    UpdateMetadata,
+)
+from chromadb.utils.embedding_functions import (
+    DefaultEmbeddingFunction,
+    known_embedding_functions,
+    register_embedding_function,
+)
+from multiprocessing import cpu_count
+import warnings
+
+
+class HNSWConfiguration(TypedDict, total=False):
+    space: Space
+    ef_construction: int
+    max_neighbors: int
+    ef_search: int
+    num_threads: int
+    batch_size: int
+    sync_threshold: int
+    resize_factor: float
+
+
+def default_hnsw_configuration() -> HNSWConfiguration:
+    return HNSWConfiguration(
+        space=Space.L2,
+        ef_construction=100,
+        max_neighbors=16,
+        ef_search=100,
+        num_threads=cpu_count(),
+        batch_size=100,
+        sync_threshold=1000,
+        resize_factor=1.2,
+    )
+
+
+def json_to_hnsw_configuration(json_map: Dict[str, Any]) -> HNSWConfiguration:
+    config: HNSWConfiguration = {}
+    if "space" in json_map:
+        space_value = json_map["space"]
+        if isinstance(space_value, str):
+            config["space"] = Space(space_value)
+        else:
+            config["space"] = space_value
+    if "ef_construction" in json_map:
+        config["ef_construction"] = json_map["ef_construction"]
+    if "max_neighbors" in json_map:
+        config["max_neighbors"] = json_map["max_neighbors"]
+    if "ef_search" in json_map:
+        config["ef_search"] = json_map["ef_search"]
+    if "num_threads" in json_map:
+        config["num_threads"] = json_map["num_threads"]
+    if "batch_size" in json_map:
+        config["batch_size"] = json_map["batch_size"]
+    if "sync_threshold" in json_map:
+        config["sync_threshold"] = json_map["sync_threshold"]
+    if "resize_factor" in json_map:
+        config["resize_factor"] = json_map["resize_factor"]
+    return config
+
+
+class CollectionConfiguration(TypedDict, total=False):
+    hnsw: Optional[HNSWConfiguration]
+    embedding_function: Optional[EmbeddingFunction[Embeddable]]
+
+
+def default_collection_configuration() -> CollectionConfiguration:
+    return CollectionConfiguration(
+        hnsw=default_hnsw_configuration(),
+        embedding_function=DefaultEmbeddingFunction(),  # type: ignore
+    )
+
+
+def load_collection_configuration_from_json_str(
+    json_str: str,
+) -> CollectionConfiguration:
+    json_map = json.loads(json_str)
+    return load_collection_configuration_from_json(json_map)
+
+
+def load_collection_configuration_from_json(
+    json_map: Dict[str, Any]
+) -> CollectionConfiguration:
+    if json_map.get("hnsw") is None:
+        if json_map.get("embedding_function") is None:
+            return CollectionConfiguration(
+                hnsw=default_create_hnsw_configuration(),
+            )
+        else:
+            ef_config = json_map["embedding_function"]
+            if ef_config["type"] == "legacy":
+                warnings.warn(
+                    "legacy embedding function config",
```

Comment:
could we make a note somewhere to make this warning prettier and link to some migration docs once we create them?